File: cli/internal/sdd/workspace.go
package sdd

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// resolveRoot picks the workspace root (workspaceRoot when set, else cwd, else
// the process working directory) and verifies it is an existing directory.
func resolveRoot(cwd, workspaceRoot string) (string, error) {
	candidate := workspaceRoot
	if strings.TrimSpace(candidate) == "" {
		candidate = cwd
	}

	root, err := absOrWorkingDir(candidate)
	if err != nil {
		return "", err
	}

	info, err := os.Stat(root)
	if err != nil {
		return "", err
	}
	if !info.IsDir() {
		return "", fmt.Errorf("workspace root is not a directory: %s", root)
	}
	return root, nil
}

// absOrWorkingDir resolves path to an absolute path, falling back to the
// process working directory when path is blank.
func absOrWorkingDir(path string) (string, error) {
	if strings.TrimSpace(path) == "" {
		return os.Getwd()
	}
	return filepath.Abs(path)
}

// listActiveChanges returns the sorted names of active changes: every direct
// subdirectory of openspec/changes/ except the reserved archive/ directory.
// A missing changes directory yields an empty list, not an error.
func listActiveChanges(root string) ([]string, error) {
	entries, err := os.ReadDir(filepath.Join(root, "openspec", "changes"))
	if err != nil {
		if os.IsNotExist(err) {
			return []string{}, nil
		}
		return nil, err
	}

	changes := make([]string, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() && entry.Name() != archiveDirName {
			changes = append(changes, entry.Name())
		}
	}
	sort.Strings(changes)
	return changes, nil
}

func contains(values []string, needle string) bool {
	for _, value := range values {
		if value == needle {
			return true
		}
	}
	return false
}

func firstPath(paths []string) string {
	if len(paths) == 0 {
		return ""
	}
	return paths[0]
}

File: cli/src/ai_harness/sdd/workspace.py
"""Workspace-root resolution and active-change discovery."""

from __future__ import annotations

import os

from .models import ARCHIVE_DIR_NAME, SddError


def resolve_root(cwd: str, workspace_root: str) -> str:
    """Pick the workspace root (workspace_root when set, else cwd, else the
    process working directory) and verify it is an existing directory."""
    candidate = workspace_root if workspace_root.strip() else cwd
    root = _abs_or_working_dir(candidate)

    if not os.path.exists(root):
        raise SddError(f"workspace root not found: {root}")
    if not os.path.isdir(root):
        raise SddError(f"workspace root is not a directory: {root}")
    return root


def _abs_or_working_dir(path: str) -> str:
    if not path.strip():
        return os.getcwd()
    return os.path.abspath(path)


def list_active_changes(root: str) -> list[str]:
    """Return the sorted names of active changes: every direct subdirectory of
    openspec/changes/ except the reserved archive/ directory. A missing changes
    directory yields an empty list, not an error."""
    changes_dir = os.path.join(root, "openspec", "changes")
    try:
        entries = os.scandir(changes_dir)
    except FileNotFoundError:
        return []
    except OSError as exc:  # pragma: no cover - surfaced as resolution failure
        raise SddError(str(exc)) from exc

    changes = []
    with entries:
        for entry in entries:
            if entry.is_dir() and entry.name != ARCHIVE_DIR_NAME:
                changes.append(entry.name)
    changes.sort()
    return changes
